File: scala/src/main/scala/part1/chapter3/Tree.scala
package part1.chapter3

enum Tree[+A]:
    case Leaf(value: A)
    case Branch(left: Tree[A], right: Tree[A])

    def size: Int =
        this match
            case Leaf(_) => 1
            case Branch(l, r) => 1 + l.size + r.size

    def depth: Int =
        this match
            case Leaf(_) => 0
            case Branch(l, r) => 1 + (l.depth.max(r.depth))

object Tree:
    def size[A](t: Tree[A]): Int =
        t match
            case Leaf(_) => 1
            case Branch(l, r) => 1 + size(l) + size(r)

    extension (t: Tree[Int]) def maximum: Int =
        t match
            case Leaf(n) => n
            case Branch(l, r) => l.maximum.max(r.maximum)

File: rust/src/part1/chapter3/tree.rs
use std::cmp::max;
use std::rc::Rc;

enum Tree<A> {
    Leaf {
        value: A,
    },
    Branch {
        left: Rc<Tree<A>>,
        right: Rc<Tree<A>>,
    },
}

impl<A> Tree<A> {
    fn size(&self) -> usize {
        match self {
            Tree::Leaf { .. } => 1,
            Tree::Branch { left, right } => 1 + left.size() + right.size(),
        }
    }

    fn depth(&self) -> usize {
        match self {
            Tree::Leaf { .. } => 0,
            Tree::Branch { left, right } => 1 + max(left.depth(), right.depth()),
        }
    }
}

impl Tree<i32> {
    fn maximum(&self) -> i32 {
        match self {
            Tree::Leaf { value } => value.clone(),
            Tree::Branch { left, right } => max(left.maximum(), right.maximum()),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_exercise325() {
        let tree = Tree::Branch {
            left: Rc::new(Tree::Branch {
                left: Rc::new(Tree::Leaf { value: 1 }),
                right: Rc::new(Tree::Leaf { value: 2 }),
            }),
            right: Rc::new(Tree::Branch {
                left: Rc::new(Tree::Leaf { value: 3 }),
                right: Rc::new(Tree::Leaf { value: 4 }),
            }),
        };
        assert_eq!(tree.maximum(), 4);
    }

    #[test]
    fn test_exercise326() {
        let tree = Tree::Branch {
            left: Rc::new(Tree::Leaf { value: 1 }),
            right: Rc::new(Tree::Branch {
                left: Rc::new(Tree::Leaf { value: 2 }),
                right: Rc::new(Tree::Branch {
                    left: Rc::new(Tree::Leaf { value: 3 }),
                    right: Rc::new(Tree::Leaf { value: 3 }),
                }),
            }),
        };
        assert_eq!(tree.depth(), 3);
    }
}
